File: frontend/src/types/transaction.ts
export type TransactionType = "income" | "expense";
export type TransactionCategory =
  | "food" | "transport" | "entertainment" | "utilities"
  | "health" | "education" | "shopping" | "savings"
  | "salary" | "freelance" | "mobile_money" | "rent" | "other";

export interface Transaction {
  id: string;
  user_id: string;
  amount: number;
  type: TransactionType;
  category: TransactionCategory;
  description: string;
  merchant: string | null;
  ai_scanned: boolean;
  ai_categorized: boolean;
  notes: string | null;
  transaction_date: string;
  created_at: string;
}

export interface TransactionSummary {
  total_income: number;
  total_expenses: number;
  balance: number;
  transaction_count: number;
  period: string;
}

export interface CreateTransactionPayload {
  amount: number;
  type: TransactionType;
  category: TransactionCategory;
  description: string;
  merchant?: string;
  notes?: string;
  transaction_date?: string;
}

File: backend/app/models/transaction.py
import uuid
from datetime import datetime
from sqlalchemy import String, Numeric, DateTime, Boolean, Text, ForeignKey, Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID
from app.db.base import Base
import enum


class TransactionType(str, enum.Enum):
    INCOME = "income"
    EXPENSE = "expense"


class TransactionCategory(str, enum.Enum):
    FOOD = "food"
    TRANSPORT = "transport"
    ENTERTAINMENT = "entertainment"
    UTILITIES = "utilities"
    HEALTH = "health"
    EDUCATION = "education"
    SHOPPING = "shopping"
    SAVINGS = "savings"
    SALARY = "salary"
    FREELANCE = "freelance"
    MOBILE_MONEY = "mobile_money"
    RENT = "rent"
    OTHER = "other"


class Transaction(Base):
    __tablename__ = "transactions"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    amount: Mapped[float] = mapped_column(Numeric(15, 2), nullable=False)
    type: Mapped[TransactionType] = mapped_column(SAEnum(TransactionType), nullable=False)
    category: Mapped[TransactionCategory] = mapped_column(SAEnum(TransactionCategory), nullable=False)
    description: Mapped[str] = mapped_column(String(500), nullable=False)
    merchant: Mapped[str | None] = mapped_column(String(255), nullable=True)
    ai_scanned: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    ai_categorized: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    receipt_url: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    transaction_date: Mapped[datetime] = mapped_column(DateTime, nullable=False, default=datetime.utcnow)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    # Relationships
    user: Mapped["User"] = relationship("User", back_populates="transactions")
